DoSomething/gambit-campaigns
# Adds autoReplyBroadcast and askChangeTopicBroadcast types [pr]

#### What's this PR do?

A first stab at splitting out our `broadcast` Contentful content type into separate content types taking on 2 of them. We'll pass these Contentful ID to a `POST /messages?origin=broadcast` request as the `broadcastId` -- but Conversations will also save the ID as the conversation topic.  These content types contain a `broadcast` field that references  `message` entries, which we use to send a broadcast message -- but also contain additional fields to return as `templates` for when the user's topic is set to one of these broadcast types.

### autoReplyBroadcast
Deprecates use of the `survey_response` topic broadcasts we currently send.

Fields:

* `name` - text field 
* `broadcast` -the outbound message to send, limited to `message` entries
* `autoReply` - auto reply message to send, also limited to `message` entries

### askChangeTopicBroadcast
Deprecates use of the `askSignup` campaign broadcasts we currently send.

Fields:

* `name`
* `topic` - This is the topic we'll change the conversation to if the user says yes to the broadcast
* `broadcast` - This is the outbound message that will be sent to user, asking them if they'd like to change current topic to this one. If topic is associated with a campaign, this will create a signup 
* `declinedChangeTopic` - Sent if user says no while in this broadcast topic
* `invalidChangeTopicResponse` - Sent if user replies with neither yes or no while in this broadcast topic
* `autoReply` - Auto reply sent after user has declined, and is now stuck in this topic (until they send a keyword or command to break out)

###  API changes

These types are available via the `GET /broadcasts` resource. A `GET /broadcasts` response has 2 new properties added:
 
* `type` - Corresponds to the broadcast content type

* `templates` - object property for when the broadcast is set to user's current topic. Empty for legacy  broadcasts.

This content type is a bit of a hybrid of both broadcast and topic, as it contains the outbound broadcast, but also contains replies for after the broadcast -- which is why it's being exposed as  a`GET /topics/:id`, and to get things working in Gambit Conversations by making the following changes:

* If the `broadcastId` sent to a `POST /messages?origin=broadcast` request is an `autoReplyBroadcast`, send the broadcast message and set the conversation.topic to the `broadcastId` sent.

* If a `POST /messages?origin=member` message is received for a user with an `autoReplyBroadcast` topic - send them the `autoReply` template.

As a sanity check while developing this, I thought a `GET /contentfulEntries/:id` endpoint would be useful - as it eliminates the guesswork of knowing which resource we're mapping a given ID to (and for an `autoReplyBroadcast`, it can show up as both ( which could be confusing 😕 ❓ I've taken a stab at updating the topics documentation accordingly).

#### How should this be reviewed?
I've created a sample `autoPostBroadcast` - 4nwTwvXmfuuYAGYgusGyyW, and will be opening up a Conversations PR to test against it.

* Verify that the `GET /broadcasts` response looks as expected for existing broadcasts, but also contains a `templates` object when type is set to `autoReplyBroadcast

 * Verify the `GET /contentfulEntries/:id` works as expected

#### Any background context you want to provide?

It'd likely make sense to return these new content types in the `GET /topics` index response as well -- but it felt like more changes than necessary at the moment, and possibly confusing. Once we start adding many broadcast entries - our cached list of topics we use for sake of finding topics by campaign quickly would turn ugly. That code in generally is ugly and should be refactored to make separate Contentful calls to filter by campaign (either by adding custom query parameters, passing query params directly to Contentful, or potentially thinking GraphQL to DRY systems)

#### Relevant tickets
https://www.pivotaltracker.com/n/projects/2092729

#### Checklist
